src/runtime/matching.ml

```diff
@@ -228,7 +228,10 @@ and collect_args env xvs ps vs =
      end in
      collect_args env xvs ps vs
 
-  | [], _::_ | _::_, [] -> assert false
+  | [], _::_ | _::_, [] ->
+     (*** XXX should we raise an error instead? A well-typed program should never get here,
+          but for an untyped one it kind of makes sense to just fail the match. *)
+     raise Match_fail
```

Comment:
We will never run untyped programs. At most, there's external's "magic", but I don't see how it makes sense to fail the match here in that case.